joint_trajectory_controller/src/joint_trajectory_controller.cpp

```diff
@@ -194,12 +196,15 @@ controller_interface::return_type JointTrajectoryController::update(
         traj_external_point_ptr_->set_point_before_trajectory_msg(
           time, state_current_, joints_angle_wraparound_);
       }
+      traj_external_point_ptr_->sample(
+        // Sample once at the beginning to establish the start time if none was given
+        time, interpolation_method_, state_desired_, start_segment_itr, end_segment_itr);
     }
 
-    // find segment for current timestamp
-    TrajectoryPointConstIter start_segment_itr, end_segment_itr;
+    // Sample setpoint for next control cycle
     const bool valid_point = traj_external_point_ptr_->sample(
-      time, interpolation_method_, state_desired_, start_segment_itr, end_segment_itr);
+      time + update_period_, interpolation_method_, state_desired_, start_segment_itr,
+      end_segment_itr);
```

Comment:
Above you are doing the initial sampling and immediately overriding with this future one. Is this intended?

Reply:
Yes. The sample call above is - as the comment says - only to establish the **actual** start time of the trajectory. However, also in the first control cycle when we started the trajectory we want to sample at $t+\Delta t$ to generate the commands.

This should be probably refactored in the trajectory representation that the start time would not implicitly be specified by the first sample, but I thought this might not be the right PR to address that...

Reply:
Thank you for the clarification